Use destroying delete for CSSValue

darinadler opened the pull request:
#### ca08a222c8067db9d4795eb81f714274df666485
<pre>
Use destroying delete for CSSValue
<a href="https://bugs.webkit.org/show_bug.cgi?id=244997">https://bugs.webkit.org/show_bug.cgi?id=244997</a>
rdar://problem/99753615

Reviewed by Yusuke Suzuki and Sam Weinig.

* Source/JavaScriptCore/runtime/TemporalPlainDate.cpp: Fixed mistake in
copyright comment.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Added CSSValuePair.h/cpp to
project file. Without this, the build didn&apos;t properly handle dependencies.

* Source/WebCore/css/CSSCrossfadeValue.cpp:
(WebCore::CSSCrossfadeValue::customTraverseSubresources const): Renamed
from traverseSubresources, wrote with logical operators instead of if statements.
* Source/WebCore/css/CSSCrossfadeValue.h: Updated for above.
* Source/WebCore/css/CSSFilterImageValue.cpp:
(WebCore::CSSFilterImageValue::customTraverseSubresources const): Ditto.
* Source/WebCore/css/CSSFilterImageValue.h: Ditto.
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
(WebCore::CSSFontFaceSrcValue::customTraverseSubresources const): Ditto.
* Source/WebCore/css/CSSFontFaceSrcValue.h: Ditto.

* Source/WebCore/css/CSSFunctionValue.cpp:
(WebCore::CSSFunctionValue::customCSSText const): Removed document argument
and moved to makeString instead of StringBuilder.
* Source/WebCore/css/CSSFunctionValue.h: Updated for above.
* Source/WebCore/css/CSSGridAutoRepeatValue.cpp:
(WebCore::CSSGridAutoRepeatValue::customCSSText const): Ditto.
* Source/WebCore/css/CSSGridIntegerRepeatValue.cpp:
(WebCore::CSSGridIntegerRepeatValue::customCSSText const): Ditto.
* Source/WebCore/css/CSSGridLineNamesValue.cpp:
(WebCore::CSSGridLineNamesValue::customCSSText const): Ditto.

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::customTraverseSubresources const): Renamed
from traverseSubresources, wrote with logical operators instead of if statements.
* Source/WebCore/css/CSSImageSetValue.h: Updated for above.
* Source/WebCore/css/CSSImageValue.cpp:
(WebCore::CSSImageValue::customTraverseSubresources const): Ditto.
* Source/WebCore/css/CSSImageValue.h: Ditto.

* Source/WebCore/css/CSSPrimitiveValue.cpp: Added global for the legacy
precision quirk. Since CSS serialization already depends on the CSS pool
global, this doesn&apos;t add any new problems, and side steps some complexity
where some customCSSText functions need document pointers and others do not.
(WebCore::CSSPrimitiveValue::customCSSText const): Removed document pointer
and use the global to figure out whether the legacy precison quirk is needed.
* Source/WebCore/css/CSSPrimitiveValue.h: Updated for above.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived): Added a function template that lets us
use the visitor pattern to implement many functions below.
(WebCore::CSSValue::visitDerived const): Ditto.
(WebCore::CSSValue::cssValueType const): Deleted. This can be done entirely
in the deprecated CSS value classes and need not be here.
(WebCore::CSSValue::traverseSubresources const): Use visitDerived and
customTraverseSubresources. In passing, &quot;accidentally&quot; fixed a bug where
CSSImageSetValue::traverseSubresources would never have been called, since
CSSImageSetValue derives from CSSValueList.
(WebCore::compareCSSValues): Deleted.
(WebCore::CSSValue::equals const): Use visitDerived. No longer need the
compareCSSValues helper.
(WebCore::CSSValue::cssText const): Removed the document argument. Use
visitDerived.
(WebCore::CSSValue::separatorCSSText): Changed to take a separator argument
so this can be shared with DeprecatedCSSOMValueList.
(WebCore::CSSValue::operator delete): Added, replacing the
destroy function. Use visitDerived, std::destroy_at and freeAfterDestruction.
(WebCore::CSSValue::destroy): Deleted.

* Source/WebCore/css/CSSValue.h: Removed Type and cssValueType. Moved the
body of the deref function out of the class definition, and made it use
delete instead of a destroy function. Removed the document argument to
cssText and made separatorCSSText protected since it&apos;s only for CSSValueList
and CSSValuePair. Replaced the destroy function with a destroying delete
implementation. Added a base version of customTraverseSubresources that
just returns false, made it private since it&apos;s only intended to be called
by the implementation of traverseSubresources.

* Source/WebCore/css/CSSValueList.cpp:
(WebCore::CSSValueList::customCSSText const): Removed document argument
and tweaked the implementation to use std::exchange instead of checking
StringBuilder::isEmpty.
(WebCore::CSSValueList::customTraverseSubresources const): Renamed
from traverseSubresources, use a modern for loop.

* Source/WebCore/css/CSSValueList.h: Updated for above changes. Also
made separator and separatorCSSText public here, leaving them protected
in the base class.

* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.h: Changed cssValueType
to return unsigned short to match the IDL file. Updated for ClassType changes.
Use Ref&lt;const&gt; instead of a const_cast.

* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
(WebCore::DeprecatedCSSOMValue::operator delete): Added, replacing the
destroy function. Use std::destroy_at and freeAfterDestruction.
(WebCore::DeprecatedCSSOMValue::cssValueType const): Updated for changes
to ClassType.
(WebCore::DeprecatedCSSOMValue::cssText const): Ditto.
(WebCore::DeprecatedCSSOMComplexValue::cssValueType const): Added.
This has the code that was previously in CSSValue::cssValueType, and
it&apos;s better to have it here, since it&apos;s specific to the deprecated CSSOM.

* Source/WebCore/css/DeprecatedCSSOMValue.h: Use unsigned short instead
of unsigned for cssValueType to match the IDL. Changed ClassType to
an enum class and gave the values short names. Use destroying delete
instead of a destroy function, which allows us to use the default deref
from RefCounted instead of having to write our own. Use Ref&lt;const&gt; instead
of a const_cast.

* Source/WebCore/css/DeprecatedCSSOMValueList.cpp:
(WebCore::DeprecatedCSSOMValueList::cssText const): Use the
CSSValueList::separatorCSSText function instead of repeating the code here
and tweaked the implementation to use std::exchange instead of checking
StringBuilder::isEmpty.

* Source/WebCore/css/DeprecatedCSSOMValueList.h: Use WTF::map
instead of CSSValueList::itemWithoutBoundsCheck.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValueInternal):
Call CSSPrimitiveValue::setUseLegacyPrecision here instead of passing
a document in.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const): Removed the document
argument that was used only for the legacy precision quirk.
* Source/WebCore/css/StyleProperties.h: Ditto.

Canonical link: <a href="https://commits.webkit.org/254846@main">https://commits.webkit.org/254846@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77617640e3a3ab2ed88f12a89d19a55c2444d9d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90451 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35037 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21083 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99787 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/157255 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94460 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33536 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28734 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82815 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96229 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96106 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26667 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77304 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26533 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81459 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69545 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82035 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34630 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15309 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76972 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32453 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16280 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26596 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3392 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36217 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39207 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/79558 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35369 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17435 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/css/CSSValue.h:
```diff
@@ -234,24 +210,14 @@ class CSSValue {
         , m_cachedCSSTextUsesLegacyPrecision(false)
+        , m_isImplicit(false)
+        , m_valueSeparator(0)
         , m_classType(classType)
```
We can make most of them as a default initializer in the members' definition.

darinadler replied:
Done

Constellation reviewed (approved):
r=me

weinig commented on Source/WebCore/css/CSSFunctionValue.cpp:
```diff
@@ -26,18 +26,11 @@
 namespace WebCore {
     
-String CSSFunctionValue::customCSSText(Document* document) const
+String CSSFunctionValue::customCSSText() const
```
Given the recursive nature, I kinda think implementing this using a StringBuilder passed in would likely be more efficient (though, that would probably require a separate function that took the StringBuilder since this this one has to return a String).

darinadler replied:
I agree that we ought to find a way to build CSS serialization that uses a `StringBuilder` rather than intermediate strings for better performance.

darinadler replied:
But I won’t make that change immediately as part of this patch, in part because I think it might hurt performance of the simplest most common cases, and I don’t want to do the work to prove that it does not.

weinig commented on Source/WebCore/css/CSSValue.cpp:
```diff
@@ -87,40 +85,124 @@ static_assert(sizeof(CSSValue) == sizeof(SameSizeAsCSSValue), "CSS value should
 DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(CSSValue);
 
-CSSValue::Type CSSValue::cssValueType() const
+template<typename Visitor> constexpr decltype(auto) CSSValue::visitDerived(Visitor&& visitor)
```
Very nice.

weinig commented on Source/WebCore/css/CSSValue.cpp:
```diff
@@ -249,264 +234,33 @@ bool CSSValue::isCSSLocalURL(StringView relativeURL)
-
-    ASSERT_NOT_REACHED();
-    return String();
+    return visitDerived([](auto& value) {
```
If you wanted to maintain the existing behavior of passing a document to some of the derived classes but still keep the nice new visitor, you could implement this like:

```c++
String CSSValue::cssText(Document* document) const
{
    return visitDerived(WTF::makeVisitor(
        [](CSSFunctionValue& value) {
             return value.customCSSText(document);
        },
        [](CSSPrimitiveValue& value) {
             return value.customCSSText(document);
        },
        [](CSSValueList& value) {
             return value.customCSSText(document);
        },
        [](auto& value) {
             return value.customCSSText();
        }
    ));
}
```

(untest code, but a pattern I have used elsewhere).

darinadler replied:
I started just like that, and ran into interesting issues given that some of the classes are derived from other of the classes and the document arguments are optional; **very** easy to write code that looks right and is wrong. For example, the version above will pass null for document all the time, given the missing `const` in the specific value reference types, and code like this doesn’t obviously do the right thing for classes derived from `CSSValueList`. I could’ve made it work, but the version I ended up with is nicer, I think, despite the global.

weinig commented on Source/WebCore/css/DeprecatedCSSOMValueList.h:
```diff
@@ -32,26 +32,26 @@ namespace WebCore {
-        for (unsigned i = 0, size = value.length(); i < size; ++i)
-            m_values.uncheckedAppend(value.itemWithoutBoundsCheck(i)->createDeprecatedCSSOMWrapper(owner));
+        m_valueSeparator = values.separator();
+        m_values.reserveInitialCapacity(values.length());
```
```c++
m_values = WTF::map(values, [] (auto& value) { return value.createDeprecatedCSSOMWrapper(owner); });
```

(or something to that nature? might need to give CSSValueList a size() function if it doesn't have one).

darinadler replied:
That’s a neat idea. Let me give it a try.

weinig commented on Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
```diff
@@ -317,16 +318,16 @@ String PropertySetCSSStyleDeclaration::getPropertyValueInternal(CSSPropertyID pr
+    if (auto* wrapper = this->wrapper()) {
+        if (auto* globalObject = wrapper->globalObject()) {
+            if (auto* document = activeDOMWindow(*globalObject).document())
+                CSSPrimitiveValue::setUseLegacyPrecision(document->quirks().needsFlightAwareSerializationQuirk());
```
Don't love this (or really any use of mutable globals these days if I am honest), too easy to become a problem down the line whether via unexpected side effects or threading. This is pretty targeted though.

darinadler replied:
Given that CSS already uses a global for the value pool, I decided that I liked this better. The old code was really inconsistent about whether the document was passed through, and the document argument that can be null was not a very good mechanism. In a In an earlier version of the patch, I had fixed things to pass the document through more consistently, and then realized it was not really valuable given how we are trying to target this quirk. Note also that this is about implementing a quirk and tying parts of the system together that normally wouldn’t be in the name of figuring out whether to apply the quirk or not. I also considered performance and you’ll see that one of the slowest things here is the code wending its way all the way from a value to the document. Also, strangely, it relies on the wrapper to do so that it only works for values that have been exposed to JavaScript. On balance, I decided that I liked this for now, and I look forward to deleting it when the single website `flightaware.com` fixes its behavior.

darinadler replied:
Still open to suggestions/advice on how to do it better though.

darinadler replied:
Another way of thinking about it is that if we find a place to put the CSS value pool rather than a global, we can put this there too.

Maybe CSS _serialization_ should involve a serializer object, and we could move this flag there along with the `StringBuilder`. I think that’s the big winner idea.

weinig commented on Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
```diff
@@ -317,16 +318,16 @@ String PropertySetCSSStyleDeclaration::getPropertyValueInternal(CSSPropertyID pr
     }
-    String value = m_propertySet->getPropertyValue(propertyID, doc);
+
+    String value = m_propertySet->getPropertyValue(propertyID);
```
auto?

darinadler replied:
Will do.

darinadler committed:
Use destroying delete for CSSValue
https://bugs.webkit.org/show_bug.cgi?id=244997
rdar://problem/99753615

Reviewed by Yusuke Suzuki and Sam Weinig.

* Source/JavaScriptCore/runtime/TemporalPlainDate.cpp: Fixed mistake in
copyright comment.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Added CSSValuePair.h/cpp to
project file. Without this, the build didn't properly handle dependencies.

* Source/WebCore/css/CSSCrossfadeValue.cpp:
(WebCore::CSSCrossfadeValue::customTraverseSubresources const): Renamed
from traverseSubresources, wrote with logical operators instead of if statements.
* Source/WebCore/css/CSSCrossfadeValue.h: Updated for above.
* Source/WebCore/css/CSSFilterImageValue.cpp:
(WebCore::CSSFilterImageValue::customTraverseSubresources const): Ditto.
* Source/WebCore/css/CSSFilterImageValue.h: Ditto.
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
(WebCore::CSSFontFaceSrcValue::customTraverseSubresources const): Ditto.
* Source/WebCore/css/CSSFontFaceSrcValue.h: Ditto.

* Source/WebCore/css/CSSFunctionValue.cpp:
(WebCore::CSSFunctionValue::customCSSText const): Removed document argument
and moved to makeString instead of StringBuilder.
* Source/WebCore/css/CSSFunctionValue.h: Updated for above.
* Source/WebCore/css/CSSGridAutoRepeatValue.cpp:
(WebCore::CSSGridAutoRepeatValue::customCSSText const): Ditto.
* Source/WebCore/css/CSSGridIntegerRepeatValue.cpp:
(WebCore::CSSGridIntegerRepeatValue::customCSSText const): Ditto.
* Source/WebCore/css/CSSGridLineNamesValue.cpp:
(WebCore::CSSGridLineNamesValue::customCSSText const): Ditto.

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::customTraverseSubresources const): Renamed
from traverseSubresources, wrote with logical operators instead of if statements.
* Source/WebCore/css/CSSImageSetValue.h: Updated for above.
* Source/WebCore/css/CSSImageValue.cpp:
(WebCore::CSSImageValue::customTraverseSubresources const): Ditto.
* Source/WebCore/css/CSSImageValue.h: Ditto.

* Source/WebCore/css/CSSPrimitiveValue.cpp: Added global for the legacy
precision quirk. Since CSS serialization already depends on the CSS pool
global, this doesn't add any new problems, and side steps some complexity
where some customCSSText functions need document pointers and others do not.
(WebCore::CSSPrimitiveValue::customCSSText const): Removed document pointer
and use the global to figure out whether the legacy precison quirk is needed.
* Source/WebCore/css/CSSPrimitiveValue.h: Updated for above.

* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived): Added a function template that lets us
use the visitor pattern to implement many functions below.
(WebCore::CSSValue::visitDerived const): Ditto.
(WebCore::CSSValue::cssValueType const): Deleted. This can be done entirely
in the deprecated CSS value classes and need not be here.
(WebCore::CSSValue::traverseSubresources const): Use visitDerived and
customTraverseSubresources. In passing, "accidentally" fixed a bug where
CSSImageSetValue::traverseSubresources would never have been called, since
CSSImageSetValue derives from CSSValueList.
(WebCore::compareCSSValues): Deleted.
(WebCore::CSSValue::equals const): Use visitDerived. No longer need the
compareCSSValues helper.
(WebCore::CSSValue::cssText const): Removed the document argument. Use
visitDerived.
(WebCore::CSSValue::separatorCSSText): Changed to take a separator argument
so this can be shared with DeprecatedCSSOMValueList.
(WebCore::CSSValue::operator delete): Added, replacing the
destroy function. Use visitDerived, std::destroy_at and freeAfterDestruction.
(WebCore::CSSValue::destroy): Deleted.

* Source/WebCore/css/CSSValue.h: Removed Type and cssValueType. Moved the
body of the deref function out of the class definition, and made it use
delete instead of a destroy function. Removed the document argument to
cssText and made separatorCSSText protected since it's only for CSSValueList
and CSSValuePair. Replaced the destroy function with a destroying delete
implementation. Added a base version of customTraverseSubresources that
just returns false, made it private since it's only intended to be called
by the implementation of traverseSubresources.

* Source/WebCore/css/CSSValueList.cpp:
(WebCore::CSSValueList::customCSSText const): Removed document argument
and tweaked the implementation to use std::exchange instead of checking
StringBuilder::isEmpty.
(WebCore::CSSValueList::customTraverseSubresources const): Renamed
from traverseSubresources, use a modern for loop.

* Source/WebCore/css/CSSValueList.h: Updated for above changes. Also
made separator and separatorCSSText public here, leaving them protected
in the base class.

* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.h: Changed cssValueType
to return unsigned short to match the IDL file. Updated for ClassType changes.
Use Ref<const> instead of a const_cast.

* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
(WebCore::DeprecatedCSSOMValue::operator delete): Added, replacing the
destroy function. Use std::destroy_at and freeAfterDestruction.
(WebCore::DeprecatedCSSOMValue::cssValueType const): Updated for changes
to ClassType.
(WebCore::DeprecatedCSSOMValue::cssText const): Ditto.
(WebCore::DeprecatedCSSOMComplexValue::cssValueType const): Added.
This has the code that was previously in CSSValue::cssValueType, and
it's better to have it here, since it's specific to the deprecated CSSOM.

* Source/WebCore/css/DeprecatedCSSOMValue.h: Use unsigned short instead
of unsigned for cssValueType to match the IDL. Changed ClassType to
an enum class and gave the values short names. Use destroying delete
instead of a destroy function, which allows us to use the default deref
from RefCounted instead of having to write our own. Use Ref<const> instead
of a const_cast.

* Source/WebCore/css/DeprecatedCSSOMValueList.cpp:
(WebCore::DeprecatedCSSOMValueList::cssText const): Use the
CSSValueList::separatorCSSText function instead of repeating the code here
and tweaked the implementation to use std::exchange instead of checking
StringBuilder::isEmpty.

* Source/WebCore/css/DeprecatedCSSOMValueList.h: Use WTF::map
instead of CSSValueList::itemWithoutBoundsCheck.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValueInternal):
Call CSSPrimitiveValue::setUseLegacyPrecision here instead of passing
a document in.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const): Removed the document
argument that was used only for the legacy precision quirk.
* Source/WebCore/css/StyleProperties.h: Ditto.

Canonical link: https://commits.webkit.org/254846@main
